### PR TITLE
Streetlights

### PR DESCRIPTION
**Pull request by Night-Pryanik**
#### Summary
SUMMARY: Content "Streetlights are now spawned on sidewalks."

#### Purpose of change
Our cities are extremely unrealistic due to the complete absence of streetlights despite we have streetlight furniture for a very long time.

#### Describe the solution
Made streetlights spawn on `t_sidewalk` terrain with vertical and horizontal distance of ~5~ 12 tiles between them. The algorithm is very simple and don't take various mapgen matters into account and thus can sometimes lead to weird situations (see the last outdated screenshot for an example). But I think it would be very hard to track all edge cases like that, so it's tolerable for now.
As for @John-Candlebury's suggestion I ensured that streetlights are spawned on the tiles adjacent to roads' pavement (at least for ordinary north-south or east-west types of roads).

#### Describe alternatives you've considered
Some other, more complex algorithm.

#### Testing
Created a new world, teleported across several cities, looked at the streets.

#### Additional context
<details> 
  <summary>Outdated screenshots with 5-tile spacing</summary>

Vertical, horizontal and curved roads with streetlights
изображение

T-shaped road intersection and cul-de-sac
изображение

Four-way intersection and dead-end
изображение

Weird mapgen I mentioned: streetlight right in front of garage entry
изображение
</details>

Cul-de-sac
изображение

Dead-end
изображение

Four-way intersection
изображение

### CODE REVIEW

**Comment by Zireael07** [comment]
5 tiles is a bit dense, it's a streetlight every 5 meters or slightly more - I'd double it.

**Comment by curstwist** [comment]
> 5 tiles is a bit dense,

**Comment by curstwist** [comment]
Sorry, miss clicked.

**Comment by John-Candlebury** [comment]
I'd try to ensure the streetlights are adjacent to the road's pavement.  That would at least make the blocked off garages somewhat usable.

**Comment by Hirmuolio** [comment]
Two or three car lengths at minimum. So 10-15 tiles. If they represent those big street lights they could be a bit futher too.
They should be far enough that they don't really bother driving as long as you pay at least some attention to them.

**Comment by Night-Pryanik** [comment]
How about 8-tile spacing?
изображение

Or 12-tile spacing?
изображение

File: src/mapgen_functions.cpp [hunk 1]
```diff
@@ -867,6 +867,17 @@ void mapgen_road( mapgendata &dat )
         }
     }
 
+    // place street lights
+    if( neighbor_sidewalks ) {
+        for( int x = 0; x < SEEX * 2 - 1; x++ ) {
+            for( int y = 0; y < SEEY * 2 - 1; y++ ) {
+                if( x % 5 == 0 && y % 5 == 0 && m->ter( point( x, y ) ) == t_sidewalk ) {
+                    m->furn_set( point( x, y ), f_street_light );
```
Comment (anothersimulacrum) [inline src/mapgen_functions.cpp:875]: Could I suggest against expanding the big list of global furniture/terrain ids and using the STATIC macro here?
```suggestion
                    m->furn_set( point( x, y ), STATIC( std::string( "f_street_light" ) ) );
```
This might also work, but I'm not sure, I'd have to review what the macro says/does.
```suggestion
                    m->furn_set( point( x, y ), STATIC( "f_street_light" ) );
```

I'd like to do this because having those ids like that makes the game throw an error when you load it without those, such as when running the game with `dda` disabled.

File: src/mapgen_functions.cpp [hunk 2]
```diff
@@ -867,6 +867,17 @@ void mapgen_road( mapgendata &dat )
         }
     }
 
+    // place street lights
+    if( neighbor_sidewalks ) {
+        for( int x = 0; x < SEEX * 2 - 1; x++ ) {
+            for( int y = 0; y < SEEY * 2 - 1; y++ ) {
+                if( x % 5 == 0 && y % 5 == 0 && m->ter( point( x, y ) ) == t_sidewalk ) {
```
Comment (8street) [inline src/mapgen_functions.cpp:874]: Why such a long IF statement?

Comment (8street) [inline src/mapgen_functions.cpp:874]:
```suggestion
        for( int x = 0; x < SEEX * 2 - 1; x += 5 ) {
            for( int y = 0; y < SEEY * 2 - 1; y += 5 ) {
                if( m->ter( point( x, y ) ) == t_sidewalk ) {
```
So much shorter.

Comment (8street) [reply src/mapgen_functions.cpp:874]: Although instead of 5, 15 is better. You need to leave space for benches and urns.

Comment (Night-Pryanik) [reply src/mapgen_functions.cpp]: Because I'm not good at algorithms at all.

Comment (Night-Pryanik) [reply src/mapgen_functions.cpp]: Used 12-tile spacing as overmap is 24 tile long, and 15 would look... not so pretty.

File: src/mapgen_functions.cpp (hunk 1 above)
Comment (Night-Pryanik) [reply src/mapgen_functions.cpp]: I'm not sure how this should work at all. I've tried both 
`m->furn_set( point( x, y ), STATIC( std::string( "f_street_light" ) )`
and
`m->furn_set( point( x, y ), STATIC( "f_street_light" ) )`
and both threw an error like `furn_set` can't use const CachedType as second argument.

Comment (anothersimulacrum) [reply src/mapgen_functions.cpp]: Probably I have misunderstood how the macro works. I will look at this more.

**Comment by Xaleth** [comment]
There are some areas in the city where there are only one or two, such as neighborhoods. Adding streetlights everywhere is also unrealistic. Kudos for adding them in :+1:

**Comment by kevingranade** [comment]
Resource https://globaldesigningcities.org/publication/global-street-design-guide/utilities-and-infrastructure/lighting-and-technology/lighting-design-guidance/
tl;dr as long as we're assuming relatively short poles, 12 meters / 12 tiles between poles works.
